### ISSUE TITLE
Code Folding

### ISSUE THREAD

**Issue by angelozerr** [eclipse-tm4e/tm4e#50]
TextMate grammar can define folding https://manual.macromates.com/en/navigation_overview

It should be cool if tm4e could provide the `org.eclipse.tm4e.ui.folding.TMFoldingStrategy` which supports this feature.

**Comment by mickaelistria** [comment]
Do you have an estimation of the ratio of textmate grammars that do enable folding? does VSCode rely on it ?

**Comment by angelozerr** [comment]
> Do you have an estimation of the ratio of textmate grammars that do enable folding? does VSCode rely on it ?

To be honnest with you, I don't remember. I must study again that but I'm very busy for the moment with JDT Codelens (an another project/topic).

**Comment by angelozerr** [comment]
@micktaelistria the only thing that I can tell you is that VSCode uses by default the indent folding strategy that I have implemented for Eclipse https://github.com/angelozerr/typescript.java/blob/master/eclipse/ts.eclipse.ide.ui/src/ts/eclipse/ide/ui/folding/IndentFoldingStrategy.java

Inside VSCode TypeScript, JSON, etc editor uses this folding strategy. I know you are working on folding feature for Generic Editor. IMHO, I think Generic Editor should use this indent folding strategy by default.

**Comment by mickaelistria** [comment]
> I think Generic Editor should use this indent folding strategy by default.


I agree. Now that Generic Editor has an extension point for reconciler and
an example about folding, it shouldn't be that hard to add it. Can you
please open an enhancement request for that?

**Comment by angelozerr** [comment]
> Can you
please open an enhancement request for that?

This bug exists https://bugs.eclipse.org/bugs/show_bug.cgi?id=512956

**Comment by mickaelistria** [comment]
> This bug exists https://bugs.eclipse.org/bugs/show_bug.cgi?id=512956

That bug is about providing the extension-point, not about providing some (default) implementation.

**Comment by angelozerr** [comment]
> That bug is about providing the extension-point, not about providing some (default) implementation.

Done https://bugs.eclipse.org/bugs/show_bug.cgi?id=520659

**Comment by angelozerr** [comment]
Today tm4e doesn't support code folding, but I tell me if it's a good idea since LSP4E supports now folding. For instance json language server supports folding. For yaml, I don't know.

**Comment by norru** [comment]
What's the status of this?

**Comment by mickaelistria** [comment]
Still open, looking for volunteers to fix it.
But as @angelozerr mentions, LSP4E now supports folding, so there is less value in using TM4E for that.

**Comment by sebthom** [comment]
It looks like some kind of folding now happens even without LSP, but it seems to be broken.

For [toml](https://github.com/oovm/vscode-toml) only two sections are marked as foldable:
![image](https://user-images.githubusercontent.com/426959/168171724-92c2afed-0dac-4fb1-851d-3b85533c245f.png)

For a [cpp](https://github.com/microsoft/vscode/tree/main/extensions/cpp) file the wrong lines are folded:
![screen](https://user-images.githubusercontent.com/426959/168172122-2c6e33bc-6fae-477d-9e1b-b66898b062ca.gif)

**Comment by cabintech** [comment]
Humm, no activity on folding support in the TM4E language configuration for a long while... would really like this but have way too little experience in Eclipse infrastructure to do it myself. It's a shame because the folding data is read from the language support file and shown in the UI - it just does not work (which is maybe worse than no support at all... spent a lot of time trying to get it to work before I found out it is actually not supported).

**Comment by sebthom** [comment]
@angelozerr @mickaelistria 
I am inclined to close this issue because implementing proper folding requires an abstract syntax tree (AST). That AST is not provided by TextMate, since parser tokens do not convey information about nesting depth in the code.  Additionally, each grammar file can freely name its tokens, so there is no consistent way to infer hierarchical structure.

Folding is best addressed via language servers, which have semantic understanding of the language.

The language configuration feature provides a mechanism called “Folding,” but it only applies to custom folding markers. Evaluation of those folding markers is already implemented in TM4E.

![Image](https://github.com/user-attachments/assets/312665a7-34d1-4858-9ab7-2b78454af0ca)

E.g. for CSS the css.language-configuration.json contains this:
```json
{
  "folding": {
    "markers": {
      "start": "^\\s*\\/\\*\\s*#region\\b\\s*(.*?)\\s*\\*\\/",
      "end": "^\\s*\\/\\*\\s*#endregion\\b.*\\*\\/"
    }
  },
}
```

Which means in CSS you can include `#region` `#endregion` comments that mark foldable regions:

![Image](https://github.com/user-attachments/assets/887a06e0-4f02-4b48-bd73-c9c388a79373)

**Comment by mickaelistria** [comment]
> Evaluation of those folding markers is already implemented in TM4E.

I think we're mostly missing a TMFoldingReconciler that would read those markers and populate the ProjectionAnnotationModel of the editor accordingly. IMO it's worth adding it, as not all LS might have folding out of the box and TM grammars often do a decent job with it. Such a reconcilier would for example allow folding for most languages in the language-pack bundle,

**Comment by mickaelistria** [comment]
> The foldings for those manual markers are already applied to the editor. 

OK, so I think that's perfect then and the issue can be marked as resolved.

**Comment by sebthom** [comment]
Apparently my memory failed me: support for `foldingStartMarker` and `foldingStopMarker` hasn't been merged into TM4E yet - it was only happily camping for half a year in a local branch on my laptop. I was “tricked” by the CSS language server, which applied the region folding I'd assumed the `TMFoldingReconciler` was handling. I now created PR #897 that implements the region folding in TM4E.
